pkg/hubble/filters/node.go

```diff
@@ -0,0 +1,134 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filters
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	flowpb "github.com/cilium/cilium/api/v1/flow"
+)
+
+// A NodeNameFilter filters on node name. An empty NodeNameFilter matches all
+// node names.
+//
+// NodeNameFilters are different to other filters as they are applied at the
+// node level, not at the individual flow level.
+//
+// Node names are hostnames optionally prefixed by a cluster name and a slash,
+// for example "k8s1" and "test-cluster/node01.company.com". Patterns match node
+// names (hostnames) and are similar to filename globs, for example "k8s*" and
+// "test-cluster/*.company.com". Literal lowercase letters, digits, hyphens,
+// dots, and forward slashes match themselves. A single star "*" matches zero or
+// more lowercase letters, digits, and hyphens (i.e. one domain name component,
+// everything except a forward slash or a dot). A double star "**" matches one
+// or more domain name componets. All other characters are invalid.
+type NodeNameFilter struct {
+	whitelistRegexp *regexp.Regexp
+	blacklistRegexp *regexp.Regexp
+}
+
+// NewNodeNameFilter returns a new NodeNameFilter with whitelist and blacklist.
+func NewNodeNameFilter(whitelist, blacklist []*flowpb.FlowFilter) (*NodeNameFilter, error) {
+	whitelistRegexp, err := compileNodeNamePatterns(whitelist)
+	if err != nil {
+		return nil, err
+	}
+	blacklistRegexp, err := compileNodeNamePatterns(blacklist)
+	if err != nil {
+		return nil, err
+	}
+
+	// short path: if there are no filters then return nil to avoid an
+	// allocation
+	if whitelistRegexp == nil && blacklistRegexp == nil {
+		return nil, nil
+	}
+
+	return &NodeNameFilter{
+		whitelistRegexp: whitelistRegexp,
+		blacklistRegexp: blacklistRegexp,
+	}, nil
+}
+
+// Match returns true if f matches nodeName.
+func (f *NodeNameFilter) Match(nodeName string) bool {
+	if f == nil {
+		return true
+	}
+	if f.whitelistRegexp != nil && !f.whitelistRegexp.MatchString(nodeName) {
+		return false
+	}
+	if f.blacklistRegexp != nil && f.blacklistRegexp.MatchString(nodeName) {
+		return false
+	}
+	return true
+}
+
+// compileNodeNamePatterns returns a regular expression equivalent to the node
+// patterns in flowFilters. If flowFilters contains no node patterns then it
+// returns nil.
+func compileNodeNamePatterns(flowFilters []*flowpb.FlowFilter) (*regexp.Regexp, error) {
+	sb := strings.Builder{}
+	sb.WriteString(`\A(?:`)
```

Comment:
Would it be preferable to use the standard `^` and `$` to denote SOL and EOL respectively? I'm not very familiar with the Go regexp implementation and had to look up why `\A` and `\z` are used here, so TIL. Are they functionally different from each other? As far as I can see, `\A`/`\z` don't support line-based matching and only operate on the full string.

Reply:
Your understanding is correct. I'm using `\A` and `\z` here as they are stricter than `^` and `$`. Node names should never have newlines in them, but we still want to make sure that we're matching on whole strings, not lines within a string.

Reply:
Alright! Doesn't look like we need to worry about that unless multiline support is enabled explicitly, though. (like `"(?m)^.*$"`) IMO it would be clearer with the standard syntax, but it's :fine: having it like this. :+1: